### PR TITLE
Add BERTTokenizer for pre-trained SentencePiece model 

### PR DESCRIPTION
**Pull request by haven-jeon**
## Description ##

#653 

- BERTTokenizer need to work with SentencePiece tokenizer model, when we build BERT from pre-training.
- Swig object(sentencepiece object) should be used in multiprocessing.

Sample code to use.

```python
# use the vocabulary from pre-trained model for tokenization
bert_tokenizer = nlp.data.BERTSPTokenizer("test-0690baed.bpe", vocabulary, lower=True)
# maximum sequence length
max_len = 128
# the labels for the two classes
all_labels = ["0", "1"]
# whether to transform the data as sentence pairs.
# for single sentence classification, set pair=False
pair = True
transform = dataset.BERTDatasetTransform(bert_tokenizer, max_len,
                                         labels=all_labels,
                                         label_dtype='int32',
                                         pad=True,
                                         pair=pair)
data_train = data_train_raw.transform(transform)
```

### Essentials ###
- [ ] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by mli** [comment]
Job PR-669/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-669/1/index.html

**Comment by haven-jeon** [comment]
> Thanks for the patch. How much does it differ from our [sentence piece tokenizer](http://gluon-nlp.mxnet.io/master/api/modules/data.html#gluonnlp.data.SentencepieceTokenizer)? (I just realized that this class wasn't shown in the data API doc.)

To use in BERTSentenceTransform I suggested this class. 
BERTSentenceTransform requires [BERTTokenizer](http://gluon-nlp.mxnet.io/master/api/modules/data.html#gluonnlp.data.BERTTokenizer) interface.

File: src/gluonnlp/data/transforms.py [hunk 1]
```diff
@@ -1011,6 +1057,95 @@ def convert_tokens_to_ids(self, tokens):
...
+        self.sentencepiece = None
+
+    def _activate_sp(self):
+        self.sentencepiece = SentencepieceTokenizer(self._path, self._num_best,
+                                                    self._alpha)
+
+    def _tokenize_wordpiece(self, text):
+        """Tokenizes a piece of text into its word pieces.
+
+        This use Google's SentencePiece tokenizer model file
```
Comment (eric-haibin-lin) [inline src/gluonnlp/data/transforms.py]: This uses

File: src/gluonnlp/data/transforms.py [hunk 2]
```diff
@@ -1011,6 +1057,95 @@ def convert_tokens_to_ids(self, tokens):
...
+            self._activate_sp()
+        output_tokens = self.sentencepiece(text)
+        return output_tokens
+
+    def convert_tokens_to_ids(self, tokens):
+        """Converts a sequence of tokens into ids using the vocab or sentencepiece model."""
+        if self.vocab is None:
+            if self.sentencepiece is None:
+                self._activate_sp()
+            return self.sentencepiece._processor.encode_as_ids(  #pylint: disable=protected-access
```
[the remark this answers is not in the document]
Comment (haven-jeon) [reply src/gluonnlp/data/transforms.py]: It needs to remove this part because `sentencepiece._processor.encode_as_ids` not applicable to work like BertVocab.

How about to add  API like `BertVocab.from_sentencepiece()`

File: src/gluonnlp/data/transforms.py (hunk 1 above)
Comment (haven-jeon) [reply src/gluonnlp/data/transforms.py]: > 
> 
> Nice work! Would you mind adding an unit test for this?

I will add.

Comment (haven-jeon) [reply src/gluonnlp/data/transforms.py]: @eric-haibin-lin How can I upload test sentencepiece model for unittest like [test_sentencepiece_tokenizer_subword_regularization](https://github.com/dmlc/gluon-nlp/blob/e2b2766a4e970affc42ff8505a69a51eafea4e19/tests/unittest/test_transforms.py#L140) ?

Comment (szha) [reply src/gluonnlp/data/transforms.py]: @haven-jeon I can help with that

Comment (haven-jeon) [reply src/gluonnlp/data/transforms.py]: @szha  Thanks. Plz let me know.

Comment (szha) [reply src/gluonnlp/data/transforms.py]: I uploaded the file to apache-mxnet/gluon/dataset/vocab/test-682b5d15.bpe

Comment (haven-jeon) [reply src/gluonnlp/data/transforms.py]: Thanks, I am working.

**Comment by mli** [comment]
Job PR-669/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-669/6/index.html

**Comment by szha** [comment]
In GluonNLP-py2-cpu-unittest:
```
_________________ test_bert_sentencepiece_sentences_transform __________________
[gw1] linux2 -- Python 2.7.15 /var/lib/jenkins/workspace/gluon-nlp-cpu-py2/conda/cpu/py2/bin/python

    @pytest.mark.remote_required
    def test_bert_sentencepiece_sentences_transform():
        url = 'http://repo.mxnet.io/gluon/dataset/vocab/test-682b5d15.bpe'
        f = download(url, overwrite=True)
        bert_vocab = BERTVocab.from_sentencepiece(f)
        bert_tokenizer = t.BERTSPTokenizer(f, bert_vocab, lower=True)
        max_len = 36
        data_train_raw = SimpleDataset(
            [['This is a very awesome, life-changing sentence.']])
        transform = t.BERTSentenceTransform(bert_tokenizer,
                                            max_len,
                                            pad=True,
                                            pair=False)
        try:
            data_train = data_train_raw.transform(transform)
        except ImportError:
            warnings.warn(
                "Sentencepiece not installed, skip test_bert_sentencepiece_sentences_transform()."
            )
            return
>       processed = list(data_train)[0]

tests/unittest/test_transforms.py:286: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
conda/cpu/py2/lib/python2.7/site-packages/mxnet/gluon/data/dataset.py:125: in __getitem__
    return self._fn(item)
src/gluonnlp/data/transforms.py:1216: in __call__
    tokens_a = self._tokenizer(text_a)
src/gluonnlp/data/transforms.py:996: in __call__
    return self._tokenizer(sample)
src/gluonnlp/data/transforms.py:1000: in _tokenizer
    for token in self.basic_tokenizer(text):
src/gluonnlp/data/transforms.py:803: in __call__
    return self._tokenize(sample)
src/gluonnlp/data/transforms.py:807: in _tokenize
    text = self._clean_text(text)
src/gluonnlp/data/transforms.py:832: in _clean_text
    if cp in (0, 0xfffd) or self._is_control(char):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <gluonnlp.data.transforms.BERTBasicTokenizer object at 0x7ff5c46e3d10>
char = 'T'

    def _is_control(self, char):
        """Checks whether `chars` is a control character."""
        # These are technically control characters but we count them as whitespace
        # characters.
        if char in ['\t', '\n', '\r']:
            return False
>       cat = unicodedata.category(char)
E       TypeError: category() argument 1 must be unicode, not str

src/gluonnlp/data/transforms.py:846: TypeError
----------------------------- Captured stdout call -----------------------------
Downloading test-682b5d15.bpe from http://repo.mxnet.io/gluon/dataset/vocab/test-682b5d15.bpe...
```

**Comment by haven-jeon** [comment]
> In GluonNLP-py2-cpu-unittest:
> 
> ```
> _________________ test_bert_sentencepiece_sentences_transform __________________
> [gw1] linux2 -- Python 2.7.15 /var/lib/jenkins/workspace/gluon-nlp-cpu-py2/conda/cpu/py2/bin/python
> 
>     @pytest.mark.remote_required
>     def test_bert_sentencepiece_sentences_transform():
>         url = 'http://repo.mxnet.io/gluon/dataset/vocab/test-682b5d15.bpe'
>         f = download(url, overwrite=True)
>         bert_vocab = BERTVocab.from_sentencepiece(f)
>         bert_tokenizer = t.BERTSPTokenizer(f, bert_vocab, lower=True)
>         max_len = 36
>         data_train_raw = SimpleDataset(
>             [['This is a very awesome, life-changing sentence.']])
>         transform = t.BERTSentenceTransform(bert_tokenizer,
>                                             max_len,
>                                             pad=True,
>                                             pair=False)
>         try:
>             data_train = data_train_raw.transform(transform)
>         except ImportError:
>             warnings.warn(
>                 "Sentencepiece not installed, skip test_bert_sentencepiece_sentences_transform()."
>             )
>             return
> >       processed = list(data_train)[0]
> 
> tests/unittest/test_transforms.py:286: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> conda/cpu/py2/lib/python2.7/site-packages/mxnet/gluon/data/dataset.py:125: in __getitem__
>     return self._fn(item)
> src/gluonnlp/data/transforms.py:1216: in __call__
>     tokens_a = self._tokenizer(text_a)
> src/gluonnlp/data/transforms.py:996: in __call__
>     return self._tokenizer(sample)
> src/gluonnlp/data/transforms.py:1000: in _tokenizer
>     for token in self.basic_tokenizer(text):
> src/gluonnlp/data/transforms.py:803: in __call__
>     return self._tokenize(sample)
> src/gluonnlp/data/transforms.py:807: in _tokenize
>     text = self._clean_text(text)
> src/gluonnlp/data/transforms.py:832: in _clean_text
>     if cp in (0, 0xfffd) or self._is_control(char):
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> self = <gluonnlp.data.transforms.BERTBasicTokenizer object at 0x7ff5c46e3d10>
> char = 'T'
> 
>     def _is_control(self, char):
>         """Checks whether `chars` is a control character."""
>         # These are technically control characters but we count them as whitespace
>         # characters.
>         if char in ['\t', '\n', '\r']:
>             return False
> >       cat = unicodedata.category(char)
> E       TypeError: category() argument 1 must be unicode, not str
> 
> src/gluonnlp/data/transforms.py:846: TypeError
> ----------------------------- Captured stdout call -----------------------------
> Downloading test-682b5d15.bpe from http://repo.mxnet.io/gluon/dataset/vocab/test-682b5d15.bpe...
> ```

Needs to add unicode convertion function can work both Py2, Py3.

File: src/gluonnlp/data/transforms.py [hunk 3]
```diff
@@ -1011,6 +1057,84 @@ def convert_tokens_to_ids(self, tokens):
...
+    num_best : int, default 0
+        A scalar for sampling subwords. If num_best = {0,1}, no sampling is performed.
+        If num_best > 1, then samples from the num_best results.
+        If num_best < 0, then assume that num_best is infinite and
+        samples from the all hypothesis (lattice) using forward-filtering-and-backward-sampling
+        algorithm.
+    alpha : float
+        A scalar for a smoothing parameter. Inverse temperature for probability rescaling.
+    lower : bool, default True
+        whether the text strips accents and convert to lower case.
```
Comment (eric-haibin-lin) [inline src/gluonnlp/data/transforms.py]: whether  -> Whether
Flase -> False

File: src/gluonnlp/vocab/bert.py [hunk 4]
```diff
@@ -201,3 +203,60 @@ def from_json(cls, json_str):
...
+        padding_token : hashable object or None, default '[PAD]'
+            The representation for padding token.
+        reserved_tokens : list of strs or None, optional
+            A list of reserved tokens that will always be indexed.
+
+        Returns
+        -------
+        BERTVocab
+        """
+        sp = SentencepieceTokenizer(path)
```
Comment (eric-haibin-lin) [inline src/gluonnlp/vocab/bert.py]: add os.path.expanduser(path)

File: src/gluonnlp/vocab/bert.py [hunk 5]
```diff
@@ -201,3 +203,60 @@ def from_json(cls, json_str):
...
+        if reserved_tokens:
+            special_tokens.extend(reserved_tokens)
+
+        #check exist special token
+        for token in special_tokens:
+            if token not in token_to_idx:
+                raise ValueError('Token `{}` is not in `{}`.'.format(
+                    token, path))
+
+        bert_vocab = cls(mask_token=mask_token,
```
Comment (eric-haibin-lin) [inline src/gluonnlp/vocab/bert.py]: shall we also pass the corresponding unk, bos, eos tokens to the constructor?

**Comment by eric-haibin-lin** [comment]
Would it work if you just change the test case to:
`[[u'This is a very awesome, life-changing sentence.']])`?

**Comment by haven-jeon** [comment]
> 
> 
> Would it work if you just change the test case to:
> `[[u'This is a very awesome, life-changing sentence.']])`?

Could you give me guide to run unite test on local?

**Comment by szha** [comment]
The steps of `GluonNLP-py2-cpu-unittest` are in https://github.com/dmlc/gluon-nlp/blob/master/ci/jenkins/Jenkinsfile_py2_cpu_unittest#L53-L56
The function called is here: https://github.com/dmlc/gluon-nlp/blob/master/ci/jenkins/build_steps.groovy#L41-L61

**Comment by eric-haibin-lin** [comment]
I see. Then maybe we need to add some util.to_unicode():
```
if py2:
    return s.decode('utf-8')
else:
    str(s)
```
and apply this to the output of sp_tokenizer.tokens?

**Comment by haven-jeon** [comment]
> 
> 
> The steps of `GluonNLP-py2-cpu-unittest` are in https://github.com/dmlc/gluon-nlp/blob/master/ci/jenkins/Jenkinsfile_py2_cpu_unittest#L53-L56
> The function called is here: https://github.com/dmlc/gluon-nlp/blob/master/ci/jenkins/build_steps.groovy#L41-L61
> 
> You will need anaconda or miniconda for creating the environment.

Thanks. 



> 
> 
> I see. Then maybe we need to add some util.to_unicode():
> 
> ```
> if py2:
>     return s.decode('utf-8')
> else:
>     str(s)
> ```
> 
> and apply this to the output of sp_tokenizer.tokens?

Yes, I'm applying unicode convertion function to vocab tokens.

**Comment by szha** [comment]
> Thanks. Does Py3 tests need other env?

Yes, there's the corresponding Jenkins file for it too.

**Comment by mli** [comment]
Job PR-669/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-669/7/index.html

**Comment by mli** [comment]
Job PR-669/8 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-669/8/index.html

**Comment by mli** [comment]
Job PR-669/9 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-669/9/index.html

**Comment by mli** [comment]
Job PR-669/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-669/10/index.html

File: src/gluonnlp/vocab/bert.py [hunk 6]
```diff
@@ -201,3 +204,70 @@ def from_json(cls, json_str):
...
+        BERTVocab
+        """
+        sp = SentencepieceTokenizer(os.path.expanduser(path))
+        token_to_idx = {
+            convert_to_unicode(t): i
+            for i, t in enumerate(sp.tokens)
+        }
+        special_tokens_dic = {
+            k: convert_to_unicode(token) if token is not None else None
+            for k, token in zip(['mask', 'sep', 'cls', 'unk', 'pad', 'bos', 'eos'], [
```
Comment (eric-haibin-lin) [inline src/gluonnlp/vocab/bert.py]: What i meant is that, the sentence piece vocab already added unk, bos, eos in its dictionary, right? You can get them by sth like
```
sp = sentencepiece()
sp.load('model')
bos_id, eos_id, unk_id =  sp.bosId(), sp.eosId(), sp.unkId()
bos_token, eos_token, unk_token = ... sp.tokens[bos_id], sp.tokens[eos_id], sp.tokens[unk_id]
```
And we should pass these bos_token, eos_token, unk_token from the sentence piece model to construct our vocab? This way the vocab is more consistent with the model from sentencepiece. Otherwise, there're '<s>' and '</s>' from sentencepiece vocab which already stand for bos, eos.

Comment (haven-jeon) [reply src/gluonnlp/vocab/bert.py]: Thanks. Your right.

**Comment by eric-haibin-lin** [comment]
@haven-jeon I made some changes to make sure the vocab is consistent with the one from the sentencepiece file. Please let me know if you think the change is reasonable. 
@leezu I need to manually construct token2idx and idx2token for the vocab due to existing inflexible special token registration in the Vocab class (special ones are always inserted in the beginning, which makes the vocab inconsistent with the original one). The code would be much more simplified once the flexible special token registration functionality is done, I guess.

**Comment by mli** [comment]
Job PR-669/12 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-669/12/index.html

**Comment by leezu** [comment]
@eric-haibin-lin thanks; I was not aware this PR is blocked by https://github.com/dmlc/gluon-nlp/issues/393.

Due to https://github.com/dmlc/gluon-nlp/blob/c80b7712e3c00b81db318981d46ea9cb49471e5b/src/gluonnlp/_constants.py#L33-L35

we need to first regenerate some of our files on S3. I'm working on it.

File: src/gluonnlp/data/transforms.py [hunk 7]
```diff
@@ -1011,6 +1057,84 @@ def convert_tokens_to_ids(self, tokens):
...
+        super(BERTSPTokenizer, self).__init__(vocab, lower,
+                                              max_input_chars_per_word)
+        self._path = path
+        self._num_best = num_best
+        self._alpha = alpha
+        self.sentencepiece = None
+
+    def _activate_sp(self):
+        self.sentencepiece = SentencepieceTokenizer(self._path, self._num_best,
+                                                    self._alpha)
```
Comment (szha) [inline src/gluonnlp/data/transforms.py]: any reason not to initialize this in constructor?

File: src/gluonnlp/data/utils.py [hunk 8]
```diff
@@ -18,26 +18,25 @@
...
-import os
 import collections
-import zipfile
+import os
 import tarfile
-import numpy as np
+import zipfile
 
+import numpy as np
+import six
```
Comment (szha) [inline src/gluonnlp/data/utils.py]: six is not yet a dependency of mxnet. any way around it?

File: src/gluonnlp/data/utils.py [hunk 9]
```diff
@@ -407,3 +406,35 @@ def __call__(self, s):
             List of strings. Obtained by calling s.split(separator).
         """
         return s.split(self._separator)
+
+
+def convert_to_unicode(text):
```
Comment (szha) [inline src/gluonnlp/data/utils.py]: `isinstance` calls on each string seems expensive. could we narrow down the interface so that it doesn't have to handle several different types?

File: src/gluonnlp/data/transforms.py (hunk 7 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/data/transforms.py]: I think the comment below explains it: 
```
#Swig object can not be pickled when multiprocessing.
```

File: src/gluonnlp/data/utils.py (hunk 9 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/data/utils.py]: Made it a private function to avoid abuse. For vocab generation the speed is fine.

File: src/gluonnlp/data/utils.py (hunk 8 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/data/utils.py]: removed

File: src/gluonnlp/data/utils.py [hunk 10]
```diff
@@ -407,3 +406,35 @@ def __call__(self, s):
...
+    text : str or bytes
+        text to be converted to unicode
+
+    Returns
+    -------
+    str
+        unicode string
+    """
+    py_version = sys.version_info[0]
+    if py_version == 3:
```
Comment (szha) [inline src/gluonnlp/data/utils.py]: python version won't change during the process, so the version check can be moved outside the function so that the check only happens once.

**Comment by mli** [comment]
Job PR-669/13 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-669/13/index.html

File: tests/unittest/test_transforms.py [hunk 11]
```diff
@@ -259,3 +261,46 @@ def test_bert_sentences_transform():
...
+                  ] * (max_len - len(token_ids))
+
+    # token ids
+    assert all(processed[0] == np.array(token_ids, dtype='int32'))
+    # sequence length
+    assert np.asscalar(processed[1]) == len(tokens) + 2
+    # segment id
+    assert all(processed[2] == np.array([0] * max_len, dtype='int32'))
+
+test_bert_sentencepiece_sentences_transform()
```
Comment (szha) [inline tests/unittest/test_transforms.py]: this call shouldn't have been added as it makes the pytest marker not function properly in some cases.

Comment (eric-haibin-lin) [reply tests/unittest/test_transforms.py]: please do
